File: DataStructures/Java/StackUsingLinkedList.java
public class StackUsingLinkedList {
    class Node {
        int data;
        Node link;
    }

    Node top;

    StackUsingLinkedList() {
        this.top = null;
    }

    public void push(int x) {
        Node temp = new Node();
        temp.data = x;
        temp.link = top;
        top = temp;
    }

    public boolean isEmpty() {
        return top == null;
    }

    public int peek() {
        if (!isEmpty()) {
            return top.data;
        } else {
            System.out.println("Stack is empty");
            return -1;
        }
    }

    public void pop() {
        if (top == null) {
            System.out.print("\nStack Underflow");
            return;
        }
        top = top.link;
    }

    public void display() {
        if (top == null) {
            System.out.printf("\nStack Underflow");
            return;
        } else {
            Node temp = top;
            while (temp != null) {
                System.out.println(temp.data);
                temp = temp.link;
            }
        }
    }

    public static void main(String[] args) {
        StackUsingLinkedList stack = new StackUsingLinkedList();
        stack.push(1);
        stack.push(2);
        stack.push(3);
        stack.push(4);
        stack.push(5);
        stack.pop();
        stack.display();
    }

}

File: DataStructures/TypeScript/StackUsingLinkedList.ts
export class StackNode {
  data: number;
  next: StackNode | null;

  constructor(data: number) {
    this.data = data;
    this.next = null;
  }
}

export class Stack {
  private top: StackNode | null;
  private capacity: number;
  private size: number;

  constructor(capacity: number) {
    this.top = null;
    this.capacity = capacity;
    this.size = 0;
  }

  // Returns the size of the Stack
  getSize(): number {
    return this.size;
  }

  // Check if the stack is empty
  isEmpty(): boolean {
    return this.getSize() === 0;
  }

  // Insert an element into the stack
  push(element: number) {
    if (this.getSize() === this.capacity)
      throw new Error("Cannot push into a full stack");

    const newNode = new StackNode(element);
    newNode.next = this.top;
    this.top = newNode;
    this.size++;
  }

  // Remove an element from the stack
  pop(): number {
    if (this.isEmpty()) throw new Error("Cannot pop from an empty stack");

    const element = this.top!.data;
    this.top = this.top!.next;
    this.size--;

    return element;
  }

  // Get the top element of the stack
  peek(): number {
    if (this.isEmpty()) throw new Error("Cannot peek into an empty stack");
    return this.top!.data;
  }

  // Print the elements of the stack
  display() {
    if (this.isEmpty()) throw new Error("Cannot display an empty stack");

    let currentNode = this.top;
    while (currentNode !== null) {
      console.log(`Current data: ${currentNode.data}`);
      currentNode = currentNode.next;
    }
  }
}
